src/appleseed/renderer/modeling/camera/camera.cpp

```diff
@@ -311,44 +368,218 @@ double Camera::extract_near_z() const
     //
 
-    // Integrate and inverse opening/closing curve y = a * x + b to transform samples.
-    float map_sample_to_linear_curve(const float a, const float b, const float x)
+    success = foundation::find_root_bisection(
+        f,
+        a,
+        b,
+        eps,
+        21, // This would be enough to get 10e-6 accuracy for the whole [a,b] interval.
+        root);
+
+    if (!success)
     {
-        float constant;
-        if (a > 0.0f)
-        {
-            // Shutter is opening.
-            constant = 0.0f;
-        }
-        else
-        {
-            // Shutter is closing.
-            constant = 1.0f - (a / 2.0f) - b;
-        }
-        return (sqrt(2.0f * a * (x - constant) + b * b) - b) / a;
+        assert(false && "Cannot inverse sample shutter curve CDF");
+        RENDERER_LOG_ERROR("Cannot inverse sample shutter curve CDF of camera \"%s\"", camera.get_path().c_str());
     }
+
+    return foundation::evaluate_bezier3(t0, t1, t2, t3, root);
 }
 
-float Camera::map_to_shutter_curve(const float sample) const
+} // namespace
+
+// Linear curve consists of multiple linear patches
+float Camera::map_to_shutter_curve_impl_linear(const float sample) const
 {
-    if (0.0f <= sample && sample < m_inverse_cdf_open_point)
+    if (sample < m_inverse_cdf_open_point)
     {
         // Shutter is opening.
-        return map_sample_to_linear_curve(m_open_linear_curve_slope, 0.0f, sample);
+        return sqrt(m_shutter_curve_linear_open_multiplier * sample);
     }
-    else if (1.0f >= sample && sample > m_inverse_cdf_close_point)
+    else if (sample > m_inverse_cdf_close_point)
     {
         // Shutter is closing.
-        return map_sample_to_linear_curve(m_close_linear_curve_slope, -m_close_linear_curve_slope, sample);
+        return 1.0f - sqrt(m_shutter_curve_linear_close_multiplier * (sample - 1.0f));
     }
     else
     {
         // Shutter is fully opened.
-        return sample / m_shutter_pdf_max_height + m_normalized_open_end_time_half;
+        return m_shutter_curve_linear_fully_opened_multiplier * sample + m_shutter_open_end_time / 2.0f;
+    }
+}
+
+// Nonlinear curve consists of two bezier curves and constant patch between them
+float Camera::map_to_shutter_curve_impl_nonlinear(const float sample) const
+{
+    if (sample < m_inverse_cdf_open_point)
+    {
+        // Sample open curve
+        return inverse_sample_curve_patch(m_shutter_curve_nonlinear_open_cdf,
+                                          sample,
+                                          0.0f,
+                                          m_shutter_curve_nonlinear_control_points_normalized[0],
+                                          m_shutter_curve_nonlinear_control_points_normalized[2],
+                                          m_normalized_open_end_time,
+                                          *this);
+    }
+    else if (sample <= m_inverse_cdf_close_point)
+    {
+        return  (sample - m_inverse_cdf_open_point) /
```

Comment:
Can we fold the division into `m_shutter_curve_nonlinear_normalization_factor`?

Reply:
You mean use multiplication by 1/m_shutter_curve_normalization_factor?

Reply:
Yeah, I would change `m_shutter_curve_nonlinear_normalization_factor` to include the 1/ part.

Reply:
Actually, it has that 1/ part. We are INVERSE sampling, so this is denormalization:) That is why we have division here.